### ISSUE TITLE
Revisit unimplemented factory constructors and default values

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#4172]
We need to introduce an extra rule about factory constructors that are subject to augmentation: The can omit default values, even for

With augmentations, a factory constructor can be unimplemented. It is possible to make a late choice about the nature of this constructor (it could be redirecting or it could be non-redirecting):

```dart
class A {
  A();
  factory A.foo(); // Unimplemented factory constructor.
}

augment class A {
  augment factory A.foo() = A;
}

class B {
  B();
  factory B.foo(); // Unimplemented factory constructor.
}

augment class B {
  augment factory B.foo() {
    return A();
  }
}
```

We may need to introduce an extra rule about these unimplemented factory constructors which says that **it is allowed for an optional parameter to omit the default value clause, even in the case where the parameter type is potentially non-nullable**.

For example:

```dart
class A {
  A([int i = 0]);
  factory A.foo([int i]); // Currently an error; this issue proposes to allow it.
}

augment class A {
  augment factory A.foo([int i]) = A; // Default value obtained from redirectee.
}
```

The general rules about augmentations state that no default value can be specified by an augmentation, it must always be specified by the introductory declaration (if there is to be a default value for that parameter at all).

This implies that the declaration above would not allow `A.foo` to be augmented with a class body (such that it is a non-redirecting factory constructor): The parameter type cannot be modified by an augmentation, and an augmentation cannot provide a default value. In other words, the fact that there is no default value has eliminated the expressive power to make the decision about the factory being redirecting or non-redirecting, it's always forced to be redirecting.

We could allow the unimplemented factory constructor to _include_ the default value and then ignore it (because it would be an error to have it) when the constructor is augmented to be redirecting:

```dart
class A {
  A([int i = 0]);
  factory A.foo([int i = 1]); // Currently allowed.
}

augment class A {
  augment factory A.foo([int i]) = A; // Currently an error; this issue suggests we might allow it.
}
```

If we do not allow this then the choice to have a default value will force the constructor to be non-redirecting.

In other words, the ability to "decide late" whether a given factory is redirecting or not will disappear as soon as there is one or more optional parameters whose type is potentially non-nullable. Of course, we can then make the declarations inconsistent: If we have two such parameters and provide a default value for exactly one of them then no augmenting declarations can be written, because they will give rise to a compile-time error no matter what.

@dart-lang/language-team, WDYT? Are you willing to allow a default value to (1) occur in an augmenting declaration, and/or (2) be declared in the introductory declaration, but then erased in the final definition because tit would be an error?

**Comment by jakemac53** [comment]
My opinion is to not change anything here other than possibly augmenting the definition of "potentially redirecting" to indicate that the presence (or absence) of a default value on an optional, non-nullable parameter will determine the type of the constructor (as in redirecting or not).

It isn't intended to be a feature that an augmentation can alter this property of the constructor, this "potentially redirecting" definition is only there to describe how we handle the case where we can't know from the introductory declaration, not to try and add flexibility.

**Comment by lrhn** [comment]
I do want to introduce flexibility. 

If a user writes a constructor signature, and as a macro to it, I do want to give the macro the flexibility to choose the implementation. 

That said, any number of details can lock the declaration onto being redirecting or not:
- `const` implies factory (for now, I have an issue asking for const non-redirecting factory constructors.)
- a default value implies non-redirecting.
- a redirection or body is obvious.

It really is just the plain signature that is ambiguous.

**Comment by eernstg** [comment]
One approach we could use would be to say that the default value can be specified in an otherwise pluripotent (sounds better than 'ambiguous' ;-) constructor declaration, and if it is an error to specify that default value in the resulting constructor definition then it will simply be ignored. We might want to maintain that it is an error if there are two conflicting default values in the same augmentation chain, or we could even allow that to enhance the overall flexibility. Example:

```dart
class A {
  A([int i = 0]);
  factory A.foo([int i = 1]); // OK.
}

augment class A {
  augment factory A.foo([int i]) = A; // OK, and the "inherited" default value is ignored.
}

class B {
  B([int i = 0]);
  factory B.foo([int i = 1]); // OK.
}

augment class B {
  augment factory B.foo([int i]) { // OK, and the default value is "inherited".
    return B(i);
  }
}

class C {
  C([int i = 0]);
  factory C.foo([int i = 1]); // OK.
}

augment class C {
  augment factory C.foo([int i = 2]) { // Error?
    return C(i);
  }
}
```

Taking one step back, I think we have the following main players when it comes to constructor augmentation:

```mermaid
graph LR
  subgraph Factory
    direction BT
    fnr["Non-redirecting<br>factory A(int i) => B(i);"] --> fp["Pluripotent<br>factory A(int i);"]
    fr["Redirecting<br>factory A(int i) = B;"] --> fp
  end
  subgraph Generative
    direction BT
    gnr["Non-redirecting<br>A(int i) {...}"] --> gp["Pluripotent<br>A(int i);"]
    gr["Redirecting<br>A(int i) : this(i + 1);"] --> gp
  end
```

This is already somewhat inhomogeneous because the pluripotent generative constructor is usable as such, but the pluripotent factory is a new invention (an 'unimplemented' factory), and it is an error unless there is an augmentation which will implement it (by redirection or by a body).

The fact that some of these arrows can be blocked by parameter default values makes the situation one more bit inhomogeneous.

**Comment by eernstg** [comment]
Perhaps we should go a step further and simply say that it is not a supported feature to change the kind of constructor using augmentation? (It's a mess!)

So we would accept an introductory declaration like `A(int i);` (generative, non-redirecting, can add a body and/or an initializer list) or a introductory declaration like `factory A(int);` (factory, non-redirecting, can add a body), but they cannot be augmented into redirecting constructors.

I think this implies that there would not be any situations where a default value can be declared in the introductory declaration, and it becomes an error to have it in the augmenting declaration, or vice versa.

This takes away a little bit of expressive power, but it also gives rise to a more consistent (and less confusing) model.

**Comment by jakemac53** [comment]
> Perhaps we should go a step further and simply say that it is not a supported feature to change the kind of constructor using augmentation? (It's a mess!)

We should in general have a principle that we don't allow you to change any which could be introspected on, because that leads to different introspection results in different phases.

And, we probably do need to give you the ability to check if a constructor is redirecting or not, because it informs how you can augment the constructor.

Therefore, I think we are left with really only two reasonable options:

- Seal the type of constructor based on its introductory declaration.
- Expose the "pluripotent" property of constructors which could be redirecting or generative. This would have to be updated in between macros and would introduce unfortunate ordering constraints and reduced parallelism.

**Comment by lrhn** [comment]
> Expose the "pluripotent" property of constructors which could be redirecting or generative.

Could be `.isNonRedirecting` and `.isRedirecting` that both return false.

(Redirecting and generative are not mutually exclusive. We have almost all eight combinations of `const` or not, factory or generative, and redirecting and non-redirecting.)

**Comment by munificent** [comment]
Now that augmentations don't need to support every possible thing a macro could want to do, I think we have the opportunity to simplify a bit here. Here's a proposal:

* A constructor like `A();` or `A() { ... }` always declares a generative non-redirecting constructor. An augmentation can provide or add a body, but can't change it to be factory or redirecting.

* A constructor like `factory A();` or `factory A() { ...}` always declares a factory non-redirecting constructor. Again, an augmentation can provide or add a body, but can't change it to be generative or redirecting.

* Redirecting constructors (generative or factory) can only be created by the initial declaration of a constructor. (The initial declaration may be inside a class augmentation, but it can't be a *constructor* augmentation.)

* Redirecting constructors can't be augmented.

In short:

* As usual, an augmentation can't change any user-visible attribute of a declaration. Once a constructor is generative/const/factory/whatever, it must stay that way.

* Redirecting constructors and augmentations don't mix.

We could allow mixing them in a future language version if really compelling. But, honestly, I would much rather go in the direction of enhanced constructors (#4246), potentially const statements (#4261), and function forwarding (#3444). I think there's a path there that would lead us to not needing redirecting constructors at all. Simplifying constructors would be *really* nice.

What do you think?

**Comment by lrhn** [comment]
I really think we should allow an "API declaration" that declares what the API properties of a member, including a constructor, should be, and let something else define the implementation.

Whether a constructor is `const` or not, or is `factory` or not, are API properties.
Whether it's redirecting is an implementation choice which is invisible in the API.

So I want to be able to write
```dart
@GenerateIt()
class A {
  A(int id);
  factory A.byName(String name);
}
```
and not lock a code generator into `A` and `A.byName` being non-redirecting.


I'd rather have the *opportunity* to allow the generator to decide the implementation strategy, and risk errors,
than just not having an option at all. It won't be an error then, unless the generator comes back and says "I won't run unless you make the constructor redirecting first". In that case, you win nothing other than having to write more text.


I prefer the variant of of augmentation where you can only add implementation once (like #4203) because it avoids a lot of ordering issues, and the entire `augmented` thing.
Then it's not an issue if one augmentation turns `A()` into a redirecting constructor and another tries to add a body, because it's just an error to try to add implementation twice.

We can also allow multiple augmentations that modify implementation, they just have to agree. If they don't, it's a compile-time error, and you should stop having two code generators modify the same thing if they can't agree.

If you end up with
```dart
class A {
  const A(int id);
  A copyWith({int? id});
}
// @<generated id="GenerateIt" v="1.1.5" t="20250213111902" h="0x25af397bc9">
class A {
  final int _id;
  final String _trace;
  const A(int id) : this._(id, "A.id($id)");
  A._(this._id, this._trace);
  A copyWith({int? id}) => A._(id ?? this._id, "$trace.copy(${id ?? ""})");
  String toString() => _trace;
}
// @</generated>
// @<generated id="other" ... >
class A {
  final int id;
  const A(this.id);
  int get hashCode => id.hashCode ^ 14f37;
  bool operator ==(Object other) => other is A && id == other.id;
}
// @</generated>
```
then we have a problem. There are two implementations of `const A(int id)`, and they don't agree, not even on whether it's redirecting or not.
Neither uses `augmented` (but you shouldn't need to).

In this case we *can* layer these on top of each other and have a reasonable result:
* First `const A(this.id)` runs to initialize `id`, then rather than calling `super`, it invokes `const A(int id) : this._(id, "...");`
* That runs and redirects to `A._(..)`.
* Which runs and initializes `_id` and `_trace` and calls `Object()`.
* When this comes back it continues `A._()` (if it hadn't been const and had a body).
* Then returns to the redirecting `A(...) : this._(...)`
* Then returns to `const A(this.id)` (if it had had a body).
* And it's done.

It's not necessarily an issue to have a non-redirecting generative constructor on top of a redirecting (it works if your super-cosntructor is redirecting too, it just takes more steps to get to the actual initializing constructor.)

It would be an issue if the generated parts were swapped, though, because a redirecting generative constructor
will not run the constructor it augments.
But that reduces the error-case to having two fragments that declare the same generative constructor, ordered such that a redirecting implementation augments a non-redirecting implementation.

Which really means that *augmenting (generative) redirecting constructors is fine*. We can do that easily.
It's augmenting *with* redirecting constructors that can be a problem, because it doesn't have a flow into the augmenting constructor.

Maybe(!) we can allow *one* redirecting constuctor too, if we think of it more like a super-constructor-invocation.
Your augmenting generative constructor fragments can include at most *one* super-constructor-invocation or 
redirection. No matter where in the stack that super/redirection occurs, it's carried to the top and performed
after all the non-redirecting constructor initialization has happened. Then it jumps to the redirecting constructor
and continues, or to the super-constructor, and when it comes back from there, it runs the bodies as normal.
Probably too high risk of running initialization for the same fragment more than once.
(I [tried this](https://gist.github.com/lrhn/4b3342ad23c7d48fec78758fdc4e71f8), it has issues. So does having redirection constructors only work on constructors in this or prior fragments.)

It's not easy. It might not be possible, but I really do want to *try* before saying that you can't augment or augment with redirecting constructors. (But I can see that they don't play well with augmentation, because they jump out of the augmentation chain.)

**Comment by rrousselGit** [comment]
Hello there!

As I read this, a few things come to mind:

1. Could non-abstract redirecting factories define default values too?
  ```dart
  factory Example([int a = 0]) = Redirecting;
  ```
  Because if not, that makes this discussion pretty much useless for Freezed (as we want the `= Redirecting`, since that's how users customise the generated class name).

And given that Freezed is the main user of redirecting factories today, that'd feel odd to me :P

2. Couldn't this be generalized to all abstract functions? Looking at you typedefs!

```dart
typedef Example = void Function([int a = 0]);
...
```


----


In terms of conflict resolution in cases of "What to do with the default / What if there's a default mismatch?"

IMO the only real answer is to completely ignore default values as part of function comparison (with _maybe_ a warning) and treat it purely as a documentation thing.

One reason is: We can already override methods with default values, with a different default:

```dart
abstract class A {
  void fn([int value = 0]);
}

class B extends A {
  // Legal
  @override
  void fn([int value = 42]) {}
}
```

**Comment by eernstg** [comment]
Here's a proposal: https://github.com/dart-lang/language/pull/4393.

**Comment by lrhn** [comment]
> Abstract methods can have default values

And then they can't be augmented.I can live with that. 

Default values is implementation, not signature.
You don't _have_ to write a default value for an abstract function declaration. It means nothing, it's just a comment, a documentation of intent. (So are names of positional parameters.)

We can also allow an abstract declaration with default values, but then an implementing augmentation shouldn't have to use the same default value. It can, it likely wants to, but nothing prevents implementing the abstract method by adding a mixin as a superclass, which has a different default value. Why bother preventing it for an augmentation.

(I do think we should do the same thing for instance and static members and for constructors.)

I think I'd go with the latter: you can write default values if you want to, but unless it's the implementing declaration, they're just comments, and have absolutely no effect on anything.

**Comment by eernstg** [comment]
If we adopt the proposal in #4499 then it will be allowed for a redirecting factory constructor to have default values. According to this proposal, they can be specified anywhere (in the introductory declaration or in an augmenting declaration). It makes no difference where they are specified, as long as each formal parameter which is able to have a default value has at most one default value specification when all the contributions are brought together (and, of course, it may be an error if at the end it has none).

This means that it is _no longer true_ that the specification of one or more default values for a parameter of an incomplete factory constructor will prevent that constructor from being implemented by redirection. It can have default values (or not) and a redirection, and this combination does not have to be an error (of course, it _can_ be an error if the types are wrong, etc). Of course, it is also possible to implement the constructor as a regular (non-redirecting) factory.

In other words, #4499 just keeps one door open which was previously closed by accident ("this factory may be incomplete, but it cannot possibly be redirecting because it has a default value"), and that's a good thing. In [this comment](https://github.com/dart-lang/language/issues/4172#issuecomment-2500100485), the incomplete factory can now have default values, and it is still possible to implement it in both ways as shown.

**Comment by eernstg** [comment]
Closing: The feature specification on augmentation has been updated in https://github.com/dart-lang/language/pull/4393 to allow the default values to be specified in an introductory or an augmenting declaration for each parameter, but at most once.
